src/off2pov.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man
.TH OFF2POV  "1" " " "off2pov: Antiprism 0.32 - http://www.antiprism.com" "User Commands"
.SH NAME
off2pov - convert OFF files to POV format
.SH SYNOPSIS
.B off2pov
[\fI\,options\/\fR] \fI\,input_files\/\fR
.SH DESCRIPTION
Convert files in OFF format to POV format for display in POV\-ray. If
input_files are not given the program reads from standard input.
.PP
Options
.HP
\fB\-h\fR,\-\-help this help message (run 'off_util \fB\-H\fR help' for general help)
.HP
\fB\-\-version\fR version information
.TP
\fB\-v\fR <rad>
radius of vertex spheres, or 'b' to have radius of balls
of the maximum size without overlap (default: ball_rad/15)
.TP
\fB\-e\fR <rad>
radius of edge cylinders (default: vertex_rad/1.5)
.TP
\fB\-V\fR <col>
default vertex colour, in form 'R,G,B,A' (3 or 4 values
0.0\-1.0, or 0\-255) or hex 'xFFFFFF' (default: 1.0,0.5,0.0)
.TP
\fB\-E\fR <col>
default edge colour, in form 'R,G,B,A' (3 or 4 values
0.0\-1.0, or 0\-255) or hex 'xFFFFFF', 'x' to hide implicit edges
(default: 0.8,0.6,0.8)
.TP
\fB\-F\fR <col>
default face colour, in form 'R,G,B,A' (3 or 4 values
0.0\-1.0, or 0\-255) or hex 'xFFFFFF' (default: 0.8,0.9,0.9)
.HP
\fB\-x\fR <elms> hide elements. The element string can include v, e and f
.IP
to hide vertices, edges and faces
.HP
\fB\-n\fR <elms> show element index number labels. The element string can
.IP
include v, e and f to label vertices, edges and faces
.HP
\fB\-s\fR <syms> show symmetry elements. The element string can include
.IP
x \- rotation axes
m \- mirror planes
r \- rotation\-reflection planes
a \- all elements (same as xmr)
.HP
\fB\-m\fR <maps> a comma separated list of colour maps used to transform colour
.IP
indexes, a part consisting of letters from v, e, f, selects
the element types to apply the map list to (default 'vef').
.HP
\fB\-t\fR <disp> select face parts to display according to winding number from:
.IP
odd, nonzero (default), positive, negative, no_triangulation
(use native polygon display)
.HP
\fB\-O\fR <type> output type, can be: 'a' all in one POV file (default),
.IP
\&'s' separate files, 'o' objects only, 't' template only
.HP
\fB\-i\fR <fils> include files (separated by commas) for every POV geometry
.HP
\fB\-j\fR <fils> include files (separated by commas) for the POV scene file
.HP
\fB\-J\fR <fils> include files (separated by commas) containing additional POV
.IP
objects for the POV scene file
.HP
\fB\-o\fR <file> write output to file (default: write to standard output)
.IP
Scene options
\fB\-D\fR <dist> distance to camera
\fB\-C\fR <cent> centre of points, in form 'X,Y,Z'
\fB\-L\fR <look> point to look at, in form 'X,Y,Z'
.IP
(default, points centre)
.TP
\fB\-R\fR <rot>
rotate about axes through centre of points, in
form 'X\-ang,Y\-ang,Z\-ang' (degrees)
.TP
\fB\-B\fR <col>
background colour, in form 'R,G,B,A' (3 or 4 values
0.0\-1.0, or 0\-255) or hex 'xFFFFFF'
.HP
\fB\-P\fR <pers> narrow the angle of perspective (range 0\-100,
.IP
default: 2, recommend 4 for stereo option \fB\-S\fR 1)
.TP
\fB\-W\fR
use lighting with shado
.HP
\fB\-S\fR <type> produce stereo output, type is 0 (default) mono, 1 stereo
.IP
with one image file, 2 stereo with two image files (use
the POV\-Ray +KFF2 option for output)
.IP
Precision options
\fB\-d\fR <dgts> number of significant digits (default 17) or if negative
.IP
then the number of digits after the decimal point
.SH "SEE ALSO"
The full documentation for
.B off2pov
is maintained as a Texinfo manual.  If the
.B info
and
.B off2pov
programs are properly installed at your site, the command
.IP
.B info off2pov
.PP
should give you access to the complete manual.


NEWS
NEWS for Antiprism: http://www.antiprism.com

Version 0.32
16-06-24 Adrian Rossiter <adrian@antiprism.com>
         New Programs
         Changes
           former internal maps now available for all programs
             names: colorful, ghart, kaplan, axes, convexity, circuits
           reverse color map feature added
           off_color_radial: now works with gray, rnd and deal maps
             ability to reverse maps
           conway: now uses canonical for planarization, better convexity
           tetra59: added report. ability for arbitrary pair. better verbosity
           n_icons: compound parts calculations now work for all n/d
           off2txt: removed (moving to antiprism_addons maintained by RK)
           kaleido: removed (moving to antiprism_addons maintained by RK)
         Fixes
           build: antiview CXXFLAGS and library linking order
           iso_delta: cases h and k parts colored in pairs
           many programs: map handling verified and color options are uniform
           minor bug fixes and improvements

Version 0.31
08-07-23 Adrian Rossiter <adrian@antiprism.com>
         New Programs
         Changes
            obj2off: now uses companion .mtl file to transfer colors to off
            build: remove debian packging directory
         Fixes
            resource models: corrected std_rd, std_rt, std_re
            off2txt: change default for -d to -16, decimals only
            symmetry: missing Cs subgroup for Oh and Cv
            off2obj: corrected obj file format
            unitile2d: correct mobius strip at join
            canonical: won't change model which is already canonical at -l n
                       won't change model but will report statistics when -n 0

Version 0.30
10-01-23 Adrian Rossiter <adrian@antiprism.com>
         New Programs
            sweep_edges: sweep edges to make a surface
         Changes
            canonical:
              -V, W coloring options added
              -F, -J convexity coloring added
              -Y to realign output
            off_color: add missing adjacent element colourings -f V, E
              and -e V
            wythoff: -C add vertex colouring, 'value' method removed, 'index'
               method colours by value. Add -m colormap option.
            conway: added element colorings -V n,s,u,d -E s,v,f,d and -F v
               added planarization options make_canonical, no planarization
               added special convexity colorings
            miller: added color options -V f,n
            stellate: added color options -V f,n
            off_color_radial: added -V color and option f
               added -E color and options e,f
            symmetro: added m3 built in color map for better number of sides
            off_normals: restored -t options to add element centers to normals
            johnson: updated j88-j90 formulas. documented other formulas
            uniform: substring search removed
            polygon: orientation changed for models 2 -s2, 2 -s4, 4, 5, 6, 8,
               construction changed for some models 5 -s3
            lat_grid: move to extra programs
            n_icons: added -b don't split bow-tie faces for twisting
               added -g make m2 longitudes (-M) invisible instead of deletion
               added -C color ... circuits with first two colors
               added -B compound part counts to -L listing options
               added -D to change d (of n/d) from default of 1
               -I: added circuit counting and comparison to theoretical values
                   added compound part calculations. hybrid compound part
                   calculations work for odd d but not yet for even d
               added -f C compound coloring with polygon
               removed -f a compound coloring by parts
            help: added alternate names and wenninger numbering sections
            build: minimum autoconf version changed to 0.69
         Fixes
            resource models: large geodesic duals corrected
            polygon: fix -L, make 1 -s2 when -a is used 
            planar/off_report: fixed winding counts algorithm

Version 0.29
01-11-21 Adrian Rossiter <adrian@antiprism.com>
         New Programs
            tetra59: generate the 59 tetrahedra with rational dihedral angles
         Changes
            to_nfold: use base colours
            wythoff: allow any path
            off_color: -e nN colour edges by number of connected faces
         Fixes
            dome_layer: honeycomb connections
            unitile2d: surface closing
            wythoff: paths not starting with index

Version 0.28
13-02-21 Adrian Rossiter <adrian@antiprism.com>
         New Programs
            poly_form: rename from minmax, rename options, add -y
              for symmetry constraint, add -a e equal-edge algorithm
              and -a p planarisation
         Changes
            to_nfold: add -x, -p to help with crossed models
            minmax: rename to poly_form
            repel: add -z for status reporting
            rotegrity: remove -T (old twist program output)
            col_util: add -w for control of color grid and (new) color bars
              -I can exclude map indexes from color grid
              -r cube as new default container in HSV/HSL mode
              -R to optionally show primary and secondary reference colors
              -U adds simple sorting of colors along with already unique sort
              -M default color system mode is now HSL
            planar:
              -V verbose lists all winding numbers found in a model
              -n map entries is now calculated
              -M default color system mode is now HSL
            wythoff: new operators and new and changed operator sequences
              (see 'wythoff -c list'), extend colour by associated element
            conway: new format for conway operators with 2 numbers e.g L_n:r
              new operators added, B,E,G,W

src/off2pov.cc
/*
   Copyright (c) 2003-2016, Adrian Rossiter

   Antiprism - http://www.antiprism.com

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the "Software"),
   to deal in the Software without restriction, including without limitation
   the rights to use, copy, modify, merge, publish, distribute, sublicense,
   and/or sell copies of the Software, and to permit persons to whom the
   Software is furnished to do so, subject to the following conditions:

      The above copyright notice and this permission notice shall be included
      in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
  IN THE SOFTWARE.
*/

/* \file off2pov.cc
   \brief Convert an OFF file to POV format
*/

#include "../base/antiprism.h"

#include <cctype>
#include <cmath>
#include <cstring>
#include <string>
#include <vector>

using std::string;
using std::vector;

using namespace anti;

class o2p_opts : public ViewOpts {
public:
  bool shadow;
  int stereo_type;
  int sig_dgts;
  char o_type;
  vector<string> scene_incs;
  vector<string> obj_incs;
  vector<string> geom_incs;

  string ofile;

  o2p_opts()
      : ViewOpts("off2pov"), shadow(false), stereo_type(-1),
        sig_dgts(DEF_SIG_DGTS), o_type('a')
  {
  }

  void process_command_line(int argc, char **argv);
  void usage();
};

void o2p_opts::usage()
{
  fprintf(stdout, R"(
Usage: %s [options] input_files

Convert files in OFF format to POV format for display in POV-ray. If
input_files are not given the program reads from standard input.

Options
%s
%s
  -O <type> output type, can be: 'a' all in one POV file (default),
            's' separate files, 'o' objects only, 't' template only
  -i <fils> include files (separated by commas) for every POV geometry
  -j <fils> include files (separated by commas) for the POV scene file
  -J <fils> include files (separated by commas) containing additional POV
            objects for the POV scene file
  -o <file> write output to file (default: write to standard output)

  Scene options
%s
  -P <pers> narrow the angle of perspective (range 0-100,
            default: 2, recommend 4 for stereo option -S 1)
  -W        use lighting with shado
  -S <type> produce stereo output, type is 0 (default) mono, 1 stereo
            with one image file, 2 stereo with two image files (use
            the POV-Ray +KFF2 option for output)

  Precision options
%s

)",
          prog_name(), help_ver_text, help_view_text, help_scene_text,
          help_prec_text);
}

void o2p_opts::process_command_line(int argc, char **argv)
{
  Status stat;
  opterr = 0;
  int c;

  handle_long_opts(argc, argv);

  while ((c = getopt(argc, argv,
                     ":hv:e:V:E:F:m:x:s:n:o:D:C:L:R:P:W:S:B:d:t:I:j:J:i:O:")) !=
         -1) {
    if (common_opts(c, optopt))
      continue;

    switch (c) {
    case 'o':
      ofile = optarg;
      break;

    case 'W':
      shadow = true;
      break;

    case 'S':
      print_status_or_exit(read_int(optarg, &stereo_type), c);
      if (stereo_type < 0 || stereo_type > 3)
        error("stereo type must be 0, 1, 2 or 3", c);
      break;

    case 'd':
      print_status_or_exit(read_int(optarg, &sig_dgts), c);
      break;

    case 'j':
      scene_incs.push_back(optarg);
      break;

    case 'J':
      obj_incs.push_back(optarg);
      break;

    case 'i':
      geom_incs.push_back(optarg);
      break;

    case 'O':
      if (strlen(optarg) != 1)
        error("output type must be exactly one character", c);
      if (!strchr("asot", *optarg))
        error("output type must be 'a' all in one, 's' separate files, 'o' "
              "objects only 't' template only",
              c);
      o_type = *optarg;
      break;

    default:
      if (!(stat = read_disp_option(c, optarg))) {
        if (stat.is_warning())
          warning(stat.msg(), c);
        else
          error(stat.msg(), c);
      }
    }
  }

  if (argc - optind >= 1)
    while (argc - optind >= 1)
      ifiles.push_back(argv[optind++]);
  else if (o_type != 't')
    ifiles.push_back("");
}

void set_geom_includes(Scene &scen, const vector<string> &includes)
{
  vector<SceneGeometry> &sc_geoms = scen.get_geoms();
  vector<SceneGeometry>::iterator sc_i;
  for (sc_i = sc_geoms.begin(); sc_i != sc_geoms.end(); ++sc_i) {
    vector<GeometryDisplay *> &disps = sc_i->get_disps();
    vector<GeometryDisplay *>::iterator disp_i;
    for (disp_i = disps.begin(); disp_i != disps.end(); ++disp_i) {
      DisplayPoly *disp = dynamic_cast<DisplayPoly *>(*disp_i);
      if (disp)
        disp->set_includes(includes);
    }
  }
}

int main(int argc, char *argv[])
{
  o2p_opts opts;
  opts.process_command_line(argc, argv);
  Scene scen = opts.scen_defs;
  opts.set_view_vals(scen);
  set_geom_includes(scen, opts.geom_incs);

  PovWriter pov;
  pov.set_o_type(opts.o_type);

  if (opts.ofile != "")
    pov.set_file_name(basename2(opts.ofile.c_str()));
  else
    pov.set_file_name("stdout");

  pov.set_includes(opts.scene_incs);
  pov.set_obj_includes(opts.obj_incs);

  if (opts.stereo_type >= 0)
    pov.set_stereo_type(opts.stereo_type);

  if (opts.shadow)
    pov.set_shadow(opts.shadow);

  FILE *ofile = stdout; // write to stdout by default
  if (opts.ofile != "") {
    ofile = fopen(opts.ofile.c_str(), "w");
    if (ofile == nullptr)
      opts.error("could not open output file \'" + opts.ofile + "\'");
  }

  pov.write(ofile, scen, opts.sig_dgts);

  if (opts.ofile != "")
    fclose(ofile);

  return 0;
}
